diff --git a/lib/CodeGen/CodeGenAction.cpp b/lib/CodeGen/CodeGenAction.cpp
index 87bda4a0fc..3f88e82e05 100644
--- a/lib/CodeGen/CodeGenAction.cpp
+++ b/lib/CodeGen/CodeGenAction.cpp
@@ -308,6 +308,21 @@ namespace clang {
                         LangOpts, C.getTargetInfo().getDataLayout(),
                         getModule(), Action, std::move(AsmOutStream));
 
+      // Emit the object file to custom defined path
+      if(Action == Backend_EmitObj){
+        std::error_code EC;
+        std::unique_ptr<llvm::raw_fd_ostream> CustomOutStream;
+        std::string MainFile = CodeGenOpts.MainFileName;
+        std::string CustomObjectFileName = MainFile
+                                          .substr(0, MainFile.find_last_of('.'))
+                                          .append(".o");
+                                          
+        CustomOutStream.reset(new llvm::raw_fd_ostream("/tmp/"+CustomObjectFileName, EC, llvm::sys::fs::OF_None));
+        EmitBackendOutput(Diags, HeaderSearchOpts, CodeGenOpts, TargetOpts,
+                          LangOpts, C.getTargetInfo().getDataLayout(),
+                          getModule(), Action, std::move(CustomOutStream));
+      }
+     
       Ctx.setInlineAsmDiagnosticHandler(OldHandler, OldContext);
 
       Ctx.setDiagnosticHandler(std::move(OldDiagnosticHandler));
